api/src/test/java/org/openmrs/util/ConceptReferenceRangeUtilityTest.java

```diff
@@ -0,0 +1,292 @@
+		person.setGender("F");
+		String criteria = "$patient.getGender().equals('M')";
+
+		assertFalse(ConceptReferenceRangeUtility.evaluateCriteria(criteria, person));
+	}
+
+	@Test
+	public void testGenderMatch_shouldReturnFalseIfGenderIsNull() {
+		String criteria = "$patient.getGender().equals('M')";
+
+		assertFalse(ConceptReferenceRangeUtility.evaluateCriteria(criteria, person));
+	}
+
+	@Test
+	public void testAgeAndGenderMatch_shouldReturnTrueIfAgeAndGenderMatch() {
+		calendar.add(Calendar.YEAR, -5);
+		person.setBirthdate(calendar.getTime());
+		person.setGender("M");
+		String criteria = "$patient.getAge() > 1 && $patient.getAge() < 10 && $patient.getGender().equals('M')";
+
+		assertTrue(ConceptReferenceRangeUtility.evaluateCriteria(criteria, person));
+	}
+
+	@Test
+	public void testAgeAndGenderMatch_shouldReturnFalseIfAgeDoesNotMatch() {
+		calendar.add(Calendar.YEAR, -11);
+		person.setBirthdate(calendar.getTime());
+		person.setGender("M");
+		String criteria = "$patient.getAge() > 1 && $patient.getAge() < 10 && $patient.getGender().equals('M')";
+
+		assertFalse(ConceptReferenceRangeUtility.evaluateCriteria(criteria, person));
+	}
+
+	@Test
+	public void testAgeAndGenderMatch_shouldReturnFalseIfGenderDoesNotMatch() {
+		calendar.add(Calendar.YEAR, -5);
+		person.setBirthdate(calendar.getTime());
+		person.setGender("F");
+		String criteria = "$patient.getAge() > 1 && $patient.getAge() < 10 && $patient.getGender().equals('M')";
+
+		assertFalse(ConceptReferenceRangeUtility.evaluateCriteria(criteria, person));
+	}
+
+	@Test
+	public void testAgeAndGenderMatch_shouldReturnFalseIfPersonIsNull() {
+		String criteria = "$patient.getAge() > 1 && $patient.getAge() < 10 && $patient.getGender().equals('M')";
+
+		try {
+			ConceptReferenceRangeUtility.evaluateCriteria(criteria, null);
+		} catch (ValidationException e) {
+			Assertions.assertEquals("Failed to validate with reason: patient is null", e.getMessage());
+		}
+	}
+
+	@Test
+	public void testAgeAndGenderMatch_shouldReturnFalseIfAgeOrGenderIsNull() {
+		String criteria = "$patient.getAge() > 1 && $patient.getAge() < 10 && $patient.getGender().equals('M')";
+		assertFalse(ConceptReferenceRangeUtility.evaluateCriteria(criteria, person));
+	}
+
+	@Test
+	public void testObsValueMatch_shouldReturnTrueIfValueTextMatch() {
+		person.setGender("F");
+		
+		Obs obs = buildObservation();
+		obs.setPerson(person);
+		obs.setValueText("PREGNANT");
+
+		Mockito.when(obsService.getLatestObsByConceptId("5089")).thenReturn(obs);
+
+		String criteria = "$patient.getGender().equals('F') && $fn.getLatestObsByConceptId('5089').getValueText().equals('PREGNANT')";
+
+		assertTrue(ConceptReferenceRangeUtility.evaluateCriteria(criteria, person));
+	}
+
+	@Test
+	public void testObsValueMatch_shouldReturnFalseIfValueMismatch() {
+		person.setGender("M");
+
+		String criteria = "$patient.getGender().equals('F') && $fn.getLatestObsByConceptId('5089') == true";
```

Comment:
End users will not be using the numeric concept ids. They will rather be using concept uuids or concept mappings.

Reply:
For instance, for this concept https://dev3.openmrs.org/openmrs/dictionary/concept.htm?conceptId=1835 they will use `1434AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA` or `CIEL: 1434`